### PR TITLE
types: merge json and types pkg

### LINKED ISSUE #37577: Merge the json and types pkg, so that the json pkg can use the `types.Time` and `types.Duration`
## Enhancement

https://github.com/pingcap/tidb/issues/9988#issuecomment-1235355500

To implement the json time and duration, the `types.Time` and `types.Duration` is needed.

### PR DESCRIPTION
**Pull request by YangKeao**
Signed-off-by: YangKeao <yangkeao@chunibyo.icu>

### What problem does this PR solve?

Issue Number: close #37577  close https://github.com/pingcap/tidb/issues/37628

### What is changed and how it works?

Simply copy every codes from the `json` pkg to the `types` pkg, and rename some of the confusing variables/functions to a better name.

Another circular dependency occurs in the `/config`, I rewrote them to a manually implementation which doesn't depend on the json pkg.

I NEED NAMESPACE :crying_cat_face:  !

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### CODE REVIEW

**Comment by hawkingrei** [comment]
Now, we use the bazel [Visibility](https://bazel.build/concepts/visibility) ability to avoid wrongly importing dependencies.

**Comment by hawkingrei** [comment]
Here is an [example](https://github.com/pingcap/tidb/pull/36634)

**Comment by xiongjiwei** [comment]
is it ready to review?

**Comment by YangKeao** [comment]
> is it ready to review?

Yes! https://github.com/pingcap/tidb/pull/37579 still needs to fix some tests, but this one is ready (and simple)

File: types/json_binary.go [hunk 1]
```diff
@@ -88,7 +88,7 @@ import (
        0x01 |   // JSON true literal
        0x02 |   // JSON false literal
 
-   number ::=  ....    // little-endian format for [u]int(16|32|64), whereas
+   number ::=  ....    // little-jsonEndian format for [u]int(16|32|64), whereas
```
Comment (xiongjiwei) [inline types/json_binary.go:91]:
```suggestion
   number ::=  ....    // little-endian format for [u]int(16|32|64), whereas
```

Comment (YangKeao) [reply types/json_binary.go:91]: Nice catch!

**Review by xiongjiwei — APPROVED** [review]

File: build/nogo_config.json [hunk 2]
```diff
@@ -200,8 +200,8 @@
       "br/pkg/lightning/mydump/": "br/pkg/lightning/mydump/",
       "br/pkg/lightning/restore/opts": "br/pkg/lightning/restore/opts",
       "executor/aggregate.go": "executor/aggregate.go",
-      "types/json/binary_functions.go": "types/json/binary_functions.go",
```
Comment (hawkingrei) [inline build/nogo_config.json:203]: If you have time make nogo enable for all the types folder. it is better.

**Review by hawkingrei — APPROVED** [review]

File: .github/workflows/integration-test-dumpling-common.yml [hunk 3]
```diff
@@ -28,6 +28,13 @@ jobs:
     steps:
       - name: 'checkout repository'
         uses: actions/checkout@v3
+        with:
+          # FIXME: the version of dumpling is getting from git
+          # if it doesn't contain a meaningful git history,
+          # the release version will be empty and the program will panic.
+          #
+          # See the `DUMPLING_LDFLAGS` in `Makefile.common` for more information
+          fetch-depth: 0
```
Comment (YangKeao) [inline .github/workflows/integration-test-dumpling-common.yml:37]: @hawkingrei PTAL, I let it download the full history of TiDB to get the version through `git describe --tags --dirty='-dev'`. I think it's necessary to pass the dumpling integration test :man_facepalming:

Comment (hawkingrei) [reply .github/workflows/integration-test-dumpling-common.yml:37]: @lance6716 PTAL

Comment (lance6716) [reply .github/workflows/integration-test-dumpling-common.yml:37]: I created an issue just a minutes ago. https://github.com/pingcap/tidb/issues/37628

LGTM

**Pull request merged**